### ISSUE TITLE
502 BAD Gateway

### ISSUE THREAD

**Issue by Mattrees007** [Bee-Mar/mmpm#40]
![Annotation 2020-08-11 140647](https://user-images.githubusercontent.com/35176916/89900714-01efc400-dbdc-11ea-931d-00a40a6ab880.png)
[mmpm-logs-2020-8-11.zip](https://github.com/Bee-Mar/mmpm/files/5057119/mmpm-logs-2020-8-11.zip)

Ive installed mmpm. no appraent issues.

the web page worked once then stopped working with the attached screenshot.

**Comment by Bee-Mar** [comment]
@Mattrees007 Please attach the log files, per the instructions on the [README](https://github.com/Bee-Mar/mmpm#look-to-the-wiki)

**Comment by Mattrees007** [comment]
[mmpm-logs-2020-8-11.zip](https://github.com/Bee-Mar/mmpm/files/5057195/mmpm-logs-2020-8-11.zip)

**Comment by Bee-Mar** [comment]
Can you also please copy and paste the command string below, then attach the generated file, which will be named `mmpm-error-output.log`
```sh
LLOG=mmpm-error-output.log; sudo systemctl status mmpm mmpm-webssh nginx >> $LOG; echo -e "\n\n--JOURNALCTL OUTPUT--\n\n" >> $LOG; journalctl -xe | grep -i mmpm >> $LOG;
```

**Comment by Mattrees007** [comment]
[mmpm-error-output.log](https://github.com/Bee-Mar/mmpm/files/5057605/mmpm-error-output.log)

as requested

**Comment by Bee-Mar** [comment]
Gotta say, I'm a little stumped right now. There's nothing in that log file that shows any fatal errors. 

Could you also get the files from NGINX? Maybe there's something in there.

```sh
zip -r mmpm-nginx-logs.zip /var/log/nginx/mmpm-*.
```

The resulting file from the command up above will be in whatever your current directory is.



Additionally, you can try making sure the ports MMPM needs are accessible by using `ufw`. It may not be installed, so the command below includes the installation method. If you're not familiar, `ufw` is a command line utility that lets you change firewall settings, like opening ports easily.

```sh
sudo apt install -y ufw
sudo ufw enable
for port_num in {7890..7893}; do sudo ufw allow $port_num; done

sudo systemctl enable nginx # just a sanity check to make sure it's enabled, but it should be already
sudo systemctl restart nginx
```

**Comment by Mattrees007** [comment]
[mmpm-access.log.2.gz](https://github.com/Bee-Mar/mmpm/files/5061574/mmpm-access.log.2.gz)

Many thanks!
logs attached

**Comment by Mattrees007** [comment]
![Annotation 2020-08-12 095131](https://user-images.githubusercontent.com/35176916/89995502-76307300-dc81-11ea-92a1-fb66fd5b51b2.png)

There was no change when adding the firewall rules

**Comment by Mattrees007** [comment]
after restarting nginx MMPM sprang into life.
![Annotation 2020-08-12 095410](https://user-images.githubusercontent.com/35176916/89995783-e212db80-dc81-11ea-8eb8-7ae3165e2755.png)

Should i be concerned by the message in ACtive modules?

**Comment by Mattrees007** [comment]
however, after i do `sudo reboot` i have to 
```
sudo systemctl enable nginx # just a sanity check to make sure it's enabled, but it should be already
sudo systemctl restart nginx
```

otherwise its "Bad Gateway" again

**Comment by Bee-Mar** [comment]
@Mattrees007 that is really weird. I'm wondering if NGINX didn't install properly. It might be worth trying to reinstall NGINX, because having to restart the service each time is very strange. Also, you should only have to run `sudo systemctl enable nginx` once, just as an FYI.

You can reinstall NGINX with `sudo apt install --reinstall nginx-full -y`

As for the warning message about the active modules on the GUI, you should follow the instructions on the Wiki for setting that up. It's pretty simple to do.

Your environment variables should be set properly if you used `mmpm --guided-setup`, so you can probably start with the **Install  the MMPM MagicMirror Module** section. You'll need to install the MMPM MagicMirror module, which can be done in the control center by clicking the bottom button on the middle panel. It's the one that has the icon with a plus sign inside the two squares.

Review this page for instructions on setting it up: https://github.com/Bee-Mar/mmpm/wiki/Status,-Hide,-Show-MagicMirror-Modules

I'll be able to take a look at the logs you posted tomorrow, and I'll get back to you if I see anything there that clues me in on what's going on.

**Comment by Bee-Mar** [comment]
@Mattrees007 I decided to glance real quick at the `mmpm-error*` log files, and noticed things like `No such file or directory` and `permission denied` when NGINX was looking for the static files. Did you by chance install `mmpm` using sudo?

**Comment by stopspazzing** [comment]
> @Mattrees007 I decided to glance real quick at the `mmpm-error*` log files, and noticed things like `No such file or directory` and `permission denied` when NGINX was looking for the static files. Did you by chance install `mmpm` using sudo?

I have same exact issue. migrated from 1.xx and got bad gateway, started purging everything and reinstalling and get 502 unless restart nginx.

Curiously I am also not able to load the modules, with 0.0.0.0 set as ip and ipwhitelist blank. I get these errors:

```
2020/08/27 21:41:31 [error] 415#415: *27 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.1.3, server: _, request: "GET / HTTP/1.1", upstream: "http://[::1]:7891/", host: "192.168.1.30:7890"
2020/08/27 21:41:31 [error] 415#415: *27 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.1.3, server: _, request: "GET /favicon.ico HTTP/1.1", upstream: "http://[::1]:7891/favicon.ico", host: "192.168.1.30:7890", referrer: "http://192.168.1.30:7890/"
2020/08/27 21:45:15 [error] 2759#2759: *5 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.1.3, server: _, request: "GET /api/mmpm/environment-vars HTTP/1.1", upstream: "http://[::1]:7891/api/mmpm/environment-vars", host: "192.168.1.30:7890", referrer: "http://192.168.1.30:7890/"
2020/08/27 21:45:15 [error] 2759#2759: *3 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.1.3, server: _, request: "GET /api/mmpm/version HTTP/1.1", upstream: "http://[::1]:7891/api/mmpm/version", host: "192.168.1.30:7890", referrer: "http://192.168.1.30:7890/"
2020/08/27 21:46:12 [error] 2759#2759: *5 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.1.3, server: _, request: "GET /api/mmpm/environment-vars HTTP/1.1", upstream: "http://[::1]:7891/api/mmpm/environment-vars", host: "192.168.1.30:7890", referrer: "http://192.168.1.30:7890/"
2020/08/27 21:47:13 [error] 2759#2759: *27 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.1.3, server: _, request: "GET / HTTP/1.1", upstream: "http://[::1]:7891/", host: "192.168.1.30:7890"
2020/08/27 21:47:31 [error] 2759#2759: *63 connect() failed (111: Connection refused) while connecting to upstream, client: 192.168.1.3, server: _, request: "GET /api/packages/marketplace HTTP/1.1", upstream: "http://[::1]:7891/api/packages/marketplace", host: "192.168.1.30:7890", referrer: "http://192.168.1.30:7890/"

```

**Comment by Bee-Mar** [comment]
@stopspazzing hmm, that is interesting. In theory having an ipv6 address set shouldn't make a difference. This could be something that I'll need to dive into MagicMirror's code base to figure out.

**Comment by ParaDescartar123** [comment]
I'm getting the same thing.

I reinstall nginx and it seems to work for a few minutes, but then it faisl again.

Then I try this:

sudo systemctl restart nginx

Same thing, works for a few minutes and then it stops working.

**Comment by Bee-Mar** [comment]
@ParaDescartar123 can you provide your log files? it might provide a clue about what's going on. You can use the CLI and run `mmpm log --zip`, or download them through the GUI in the control center

**Comment by ParaDescartar123** [comment]
> @ParaDescartar123 can you provide your log files? it might provide a clue about what's going on. You can use the CLI and run `mmpm log --zip`, or download them through the GUI in the control center

Log files attached, but it seems to be more stable and hasn't stopped working for the last 5 hours after I did this:

sudo systemctl restart nginx

[mmpm-logs-2020-8-5.zip](https://github.com/Bee-Mar/mmpm/files/5177414/mmpm-logs-2020-8-5.zip)

**Comment by Bee-Mar** [comment]
@ParaDescartar123 sorry, forgot to also ask for the NGINX logs, can can you attach the zip from created from the command below:

`zip -r mmpm-nginx-logs.zip /var/log/nginx/mmpm-{access,error}.log*`

**Comment by pyrosmiley** [comment]
Hopping on here because I'm running into the same issue. have to restart the `nginx` daemon before it works. Fortunately, I just installed last night so I was able to go back through and confirm that I did _not_ install mmpm with `sudo`. I'll attach the mmpm and nginx logs for you. Let me know if there's anything else I can try!
[mmpm-nginx-logs.zip](https://github.com/Bee-Mar/mmpm/files/5178768/mmpm-nginx-logs.zip)
[mmpm-logs-2020-8-6.zip](https://github.com/Bee-Mar/mmpm/files/5178769/mmpm-logs-2020-8-6.zip)

**Comment by Bee-Mar** [comment]
@pyrosmiley just curious, do you have pi-hole or anything like that installed on the same device?

**Comment by pyrosmiley** [comment]
> @pyrosmiley just curious, do you have pi-hole or anything like that installed on the same device?

Nope, it's a fresh install of base raspbian only running the mirror software. I installed `mmpm` first since I thought it acted as an install method for the base framework, but then used [sdetweil's install script]( https://github.com/sdetweil/MagicMirror_scripts). The one nonstandard thing going on with the process is that the install script didn't properly add MM to pm2 so I had to do that manually. 

Right now things are running pretty smoothly now that I added a command to restart nginx to LXDE's autostart, but this seems kludgy.

**Comment by gnurph69** [comment]
Couple of comments, not sure if this is the same problem - but I doubt it'll hurt.

New install, manual install of MM, running PM2.  I have the same issue with Active Modules not showing up; MM running at 192.168.0.100; I disabled IPv6 (because the log looked like it was trying to retrieve an IPv6 address), no impact.  If I pass the rest of the URL (/api/packages/installed) - I get a JSON-looking response.  Haven't run into the nginx problem, but I'm rebooting every 10 minutes as I troubleshoot, so...who knows.  I'm trying to access the gui interface from a different machine than the MM machine - it occurs to me that might be the problem?

mmpm env
{
  "MMPM_MAGICMIRROR_ROOT": "/home/pi/MagicMirror",
  "MMPM_MAGICMIRROR_URI": "http://192.168.0.100:8080",
  "MMPM_MAGICMIRROR_PM2_PROCESS_NAME": "mm",
  "MMPM_MAGICMIRROR_DOCKER_COMPOSE_FILE": "",
  "MMPM_IS_DOCKER_IMAGE": false
}

what's strange, though, is this:

mmpm list -g
http://127.0.1.1:7890

Shed any light on things?

**Comment by ModestoHagney** [comment]
I'm also having the same sort of issues.

Decided to do a fresh install on a 3B+ used the latest Raspberry Pi OS, installed MMPM and installed MagicMirror from that. After a reboot I can't access the GUI until I restart nginx and I used to be able to access the site the mirror displays from another device and now i can't even do that.

I also have the same issues with MMPM gui unable to retrieve active modules. I'm pretty sure I've followed the guide to the letter but I probably ballsed something up I'm sure. I may give it a go with an older version of Raspbian, and see how it goes.

**Comment by Bee-Mar** [comment]
@pyrosmiley @ModestoHagney @gnurph69 I'll be looking into this some more later today. This has been hard to pin down as well, but i'll keep working to find a solution

**Comment by ModestoHagney** [comment]
Having another play around tonight I got this when I reinstalled MMPM on a fresh install:


`The script gunicorn is installed in '/home/pi/.local/bin' which is not on PATH.
Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  
The script flask is installed in '/home/pi/.local/bin' which is not on PATH.
Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.

The script wssh is installed in '/home/pi/.local/bin' which is not on PATH.
Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.

The script mmpm is installed in '/home/pi/.local/bin' which is not on PATH.
Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.`

I added this directory to PATH and it didn't help anything. Not sure if it's any help, but I thought I'd add it in here.


Someone further up was having the same issue as me with the GUI not being able to retrieve active modules, I got mine working. I had to change the `address:` variable in my config file to the IP of the pi and it worked first pop. Not sure if this was in the documentation and I just missed it or if it's meant to be left as 'localhost' but maybe the MagicMirror config file that came with the install through MMPM had 'localhost' instead of something else, I haven't done any testing but I presume '0.0.0.0' would also work in this situation.

**Comment by mdeuerlein** [comment]
I had the same problems but I think and for me it looks like this problem maybe somehow depends 
in which order MM and MMPM or their nginx-instances are started/initialized. e.g.

1. Startup and boot system
  => MM is working but MMPM shows Error 502
2. `sudo systemctl restart nginx`
  => now both are working (also after a few hours)
3. `pm2 restart MagicMirror` or 
  => both are still working
4. `sudo systemctl restart mm`
  => both are still working
5. `sudo systemctl restart nginx`
  => both are still working

**Comment by umbynos** [comment]
I've got the same 502 problems after a reboot. Does someone succeed in resolving?

**Comment by Bee-Mar** [comment]
@umbynos @mdeuerlein @pyrosmiley @ModestoHagney if you guys run `which gunicorn` what do you get as output?

**Comment by umbynos** [comment]
Hi! By running `which gunicorn` i get `/home/pi/.local/bin/gunicorn`

**Comment by Bee-Mar** [comment]
@umbynos @Mattrees007 @pyrosmiley @mdeuerlein this will be resolved shortly. The issue is caused by the version of `eventlet` in mmpm 2.09. The next version will be using a newer version of eventlet and a couple other bug fixes.

**Comment by Bee-Mar** [comment]
this should be solved with the latest version of mmpm just released today (2.11)

**Comment by jkriegshauser** [comment]
Btw, I ran into this issue today on version 2.11. The error logs show that there was a failure to connect to upstream but used the IPV6 address `[::1]`. On my RPi in _/etc/hosts_ I noticed that the name `localhost` was used for both `127.0.0.1` and `[::1]`. I first tried  removing `localhost` as a name for `[::1]` in the _/etc/hosts_ file, but then I would just get connection refused. So then I changed the references to `localhost` in _/etc/nginx/sites-enabled/mmpm.conf_ to `127.0.0.1` and that fixed the Bad Gateway issue without needing a nginx restart after boot.
